# The following Python Notebook will allow the user to use a bounding box to retrieve the elevation of each point in an existing array within a bounding box.
[Link to StackOverflow question](https://stackoverflow.com/questions/26000336/execute-curl-command-within-a-python-script) on using HTTP requests in Python.
[Link to documentation](https://requests.readthedocs.io/en/latest/) on Python requests library.

# Basics of 2D arrays in Python
The following section is going to design basic 2d arrays for informational purposes. It's something I have never done before. So we test it first.

In [1]:
# Print a 2d Array in python. Two Methods.
import random
array = [[ random.randint(0,4) for x in range(0,4)] for y in range(0,4)]
print(array)

two_d_array = [[2,4,8],[1,5,8],[9,3,7]]
print(two_d_array)

[[0, 2, 0, 3], [1, 1, 3, 1], [4, 1, 3, 0], [4, 1, 0, 2]]
[[2, 4, 8], [1, 5, 8], [9, 3, 7]]


# Basics of Python Request Library (Running basic requests)
## This section is going to describe how to query the API to return the Elevation data for the surrounding area.

In [2]:
import requests

# Run a GET request
r = requests.get('https://api.github.com/events')

# Display the output in JSON format
r.json()

[{'id': '43168584292',
  'type': 'PushEvent',
  'actor': {'id': 41898282,
   'login': 'github-actions[bot]',
   'display_login': 'github-actions',
   'gravatar_id': '',
   'url': 'https://api.github.com/users/github-actions[bot]',
   'avatar_url': 'https://avatars.githubusercontent.com/u/41898282?'},
  'repo': {'id': 843909763,
   'name': 'falcacer/falcacer',
   'url': 'https://api.github.com/repos/falcacer/falcacer'},
  'payload': {'repository_id': 843909763,
   'push_id': 20875804436,
   'size': 1,
   'distinct_size': 1,
   'ref': 'refs/heads/output',
   'head': 'bf1b67623348204132357c03752596e60c674def',
   'before': 'b4f10e6a9c40c505f31ea4a6dc110b0c7f1db73b',
   'commits': [{'sha': 'bf1b67623348204132357c03752596e60c674def',
     'author': {'email': '41898282+github-actions[bot]@users.noreply.github.com',
      'name': 'github-actions[bot]'},
     'message': 'Update snake animations',
     'distinct': True,
     'url': 'https://api.github.com/repos/falcacer/falcacer/commits/bf1b676

## This section will run Googles test data for the API.
After researching the given location myself, it turns out this location is the Colorado State Capitol building, which rests at an elevation of about 1609 metres. Which is then returned by the API upon request.

In [ ]:
import requests
# Sample Given by Google API documentation. *Note, insert an API key to run this request. It has been removed for security reasons.
r = requests.get('https://maps.googleapis.com/maps/api/elevation/json?locations=39.7391536%2C-104.9847034&key=api_key')
r.json()
# Returns the elevation of the Denver Capitol Building.

## The following is the result of plugging the co-ordinates into Google maps for the summit of Carrauntoohil.
Using Google Maps, the summit of Carrauntoohil returns coordinates of 51.999445 for latitude, and -9.742693 for the latitude.
Using these coordinates, the returned result should roughly match Carrauntoohils reported summit height of 1038.6

In [1]:
import requests
# Note: The API code has been removed from this for security concerns. It would have to be reinserted to work again. It worked, you'll just have to take my word for it.
res = requests.get('https://maps.googleapis.com/maps/api/elevation/json?locations=51.999445%2C-9.742693&key=apikey')
res.json()
# This has been run twice. The one above has been run once
api_call_counter = 5

{'results': [{'elevation': 994.4034423828125,
   'location': {'lat': 51.999445, 'lng': -9.742693},
   'resolution': 152.7032318115234}],
 'status': 'OK'}

## The results from the previous Python code is the following:
```
{
    "results": [
        {
            "elevation": 994.4034423828125,
            "location": {
                "lat": 51.999445,
                "lng": -9.742693
            },
            "resolution": 152.7032318115234
        }
    ],
    "status": "OK"
}
```

Where the Elevation property returns 994 meters. This is close enough to the summit, although I thought it would have been closer, given this is near the Google maps area for the summit,
and it is what [Geohack](https://geohack.toolforge.org/geohack.php?pagename=Carrauntoohil&params=51.999445_N_9.742693_W_type:mountain_scale:100000) reports with their page on the summit.

## The next objective is to draw a distance from the summit of exactly 1 kilometre for the bounding box, then find the coordinates for the box.
Given that the Summit is at 51.999445, -9.742693 the task here is to calculate the coordinates that are 1 km away in any direction.

In [4]:
# The following block of code will design a 2d array of bounding box co-ordinates
bounding_box=[[52.00844, -9.7573],[52.00844,-9.728086],[51.99045,-9.7573],[51.99045,-9.728086]]
print(bounding_box)

max_latitude = bounding_box[0][0]
min_latitude = bounding_box[2][0]
latitude_difference = max_latitude - min_latitude
print(round(latitude_difference, 5)) # Answer is corrected to 5 decimal places to prevent rounding errors.

max_longitude = bounding_box[0][1]
min_longitude = bounding_box[1][1]
longitude_difference = max_longitude - min_longitude
print(round(longitude_difference, 6))

[[52.00844, -9.7573], [52.00844, -9.728086], [51.99045, -9.7573], [51.99045, -9.728086]]
0.01799
-0.029214


## Designing a for loop to make api calls with the required co-ordinates.

In [6]:
steps = 20 # Step counter has to go to 19 instead of 20 to prevent loop index errors. Previously, it went up to 21 each.

# Variables to store the difference in Latitude and Longitude in max and minumum value, etc.
lat_diff = (max_latitude - min_latitude) / steps
long_diff = (max_longitude - min_longitude) / steps

# I was going to keep a counter until I calculated the number of requests needed and its less than the maximum number of requests available.
counter = 0
for i in range(0,steps + 1):
    for j in range (0,steps + 1):
        counter += 1
        #lat = min_latitude + i * lat_diff
        #lng = min_longitude + i * long_diff
        print("Coords (" + str(round(min_latitude + i * lat_diff, 7)) + ", " + str(round(min_longitude + j * long_diff, 7)) + ")")
print(counter)

Coords (51.99045, -9.728086)
Coords (51.99045, -9.7295467)
Coords (51.99045, -9.7310074)
Coords (51.99045, -9.7324681)
Coords (51.99045, -9.7339288)
Coords (51.99045, -9.7353895)
Coords (51.99045, -9.7368502)
Coords (51.99045, -9.7383109)
Coords (51.99045, -9.7397716)
Coords (51.99045, -9.7412323)
Coords (51.99045, -9.742693)
Coords (51.99045, -9.7441537)
Coords (51.99045, -9.7456144)
Coords (51.99045, -9.7470751)
Coords (51.99045, -9.7485358)
Coords (51.99045, -9.7499965)
Coords (51.99045, -9.7514572)
Coords (51.99045, -9.7529179)
Coords (51.99045, -9.7543786)
Coords (51.99045, -9.7558393)
Coords (51.99045, -9.7573)
Coords (51.9913495, -9.728086)
Coords (51.9913495, -9.7295467)
Coords (51.9913495, -9.7310074)
Coords (51.9913495, -9.7324681)
Coords (51.9913495, -9.7339288)
Coords (51.9913495, -9.7353895)
Coords (51.9913495, -9.7368502)
Coords (51.9913495, -9.7383109)
Coords (51.9913495, -9.7397716)
Coords (51.9913495, -9.7412323)
Coords (51.9913495, -9.742693)
Coords (51.9913495, -9.74

## Practicing with Requests
The following script will return 

In [7]:
import json
def practice_run(num):
    url = 'http://admin:mtu12345@localhost:5984/project/' + num
    response = requests.get(url)
    response_as_json = response.json()
    json_str = json.dumps(response_as_json)
    resp = json.loads(json_str)
    #f = open("elevations.txt", 'a')
    print(resp['results'][0]['location']['lat'],",", resp['results'][0]['location']['lng'],",", round(resp['results'][0]['elevation'], 3), sep='') # This produces the correct string. sep='' suppresses space after data point.
    #f.write(resp['results'][0]['location']['lat'],",", resp['results'][0]['location']['lng'],",", round(resp['results'][0]['elevation'], 3), sep='')

practice_run('5786b3f1149f34b06fac9b3e0900102e')

ConnectionError: HTTPConnectionPool(host='localhost', port=5984): Max retries exceeded with url: /project/5786b3f1149f34b06fac9b3e0900102e (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001DF86788590>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))

# The finished script
This cell contains the complete script, ready to use.

In [17]:
import json
import requests
import time

bounding_box=[[52.00844, -9.7573],[52.00844,-9.728086],[51.99045,-9.7573],[51.99045,-9.728086]]
print(bounding_box)

max_latitude = bounding_box[0][0]
min_latitude = bounding_box[2][0]
latitude_difference = max_latitude - min_latitude

max_longitude = bounding_box[0][1]
min_longitude = bounding_box[1][1]
longitude_difference = max_longitude - min_longitude

steps = 20 # Step counter has to go to 19 instead of 20 to prevent loop index errors. Previously, it went up to 21 each.

# Variables to store the difference in Latitude and Longitude in max and minumum value, etc.
lat_diff = (max_latitude - min_latitude) / steps
long_diff = (max_longitude - min_longitude) / steps

steps = 20

def return_elevation(lat, lng):
    url = "https://maps.googleapis.com/maps/api/elevation/json?locations=" + str(lat) + "%2C" + str(lng) + "&key=apikey"
    response = requests.get(url)
    response_as_json = response.json()
    json_str = json.dumps(response_as_json)
    resp = json.loads(json_str)
    lat = resp['results'][0]['location']['lat']
    lng = resp['results'][0]['location']['lng']
    elev = resp['results'][0]['elevation']
    print(lat, lng, elev)

for i in range(0,steps + 1):
    for j in range (0,steps + 1):
        return_elevation(min_latitude + i * lat_diff, min_longitude + j * long_diff)

[[52.00844, -9.7573], [52.00844, -9.728086], [51.99045, -9.7573], [51.99045, -9.728086]]
51.99045 -9.728086 814.79931640625
51.99045 -9.7295467 819.7587890625
51.99045 -9.7310074 819.8703002929688
51.99045 -9.7324681 800.9075927734375
51.99045 -9.7339288 757.3310546875
51.99045 -9.7353895 699.0140380859375
51.99045 -9.7368502 645.3720092773438
51.99045 -9.7383109 591.9722290039062
51.99045 -9.7397716 532.7890625
51.99045 -9.7412323 483.2340393066406
51.99045 -9.742693 456.0072021484375
51.99045 -9.7441537 434.3227844238281
51.99045 -9.745614400000001 415.1820373535156
51.99045 -9.7470751 421.8959655761719
51.99045 -9.7485358 446.8207092285156
51.99045 -9.7499965 473.3131408691406
51.99045 -9.7514572 509.63427734375
51.99045 -9.7529179 552.0576782226562
51.99045 -9.7543786 590.045166015625
51.99045 -9.7558393 624.4656372070312
51.99045 -9.7573 655.9550170898438
51.99134950000001 -9.728086 786.8492431640625
51.99134950000001 -9.7295467 777.4510498046875
51.99134950000001 -9.7310074 758.3